nav2_util/include/nav2_util/simple_action_server.hpp

```diff
@@ -97,8 +97,29 @@ class SimpleActionServer
   }
 
   rclcpp_action::CancelResponse handle_cancel(
-    const std::shared_ptr<rclcpp_action::ServerGoalHandle<ActionT>>/*handle*/)
+    const std::shared_ptr<rclcpp_action::ServerGoalHandle<ActionT>> handle)
   {
+    if (!handle->is_active()) {
```

Comment:
Here's a question: What if we kept this `if (!handle->is_active()) { ... }` condition, but placed it after the `std::lock_guard<std::recursive_mutex> lock(update_mutex_);`? 

If `succeeded_current` is called first, then the `update_mutex_` would be locked, and once it was done, then `handle` would no longer be active (or valid) and then it could use this condition to `REJECT` without the need of an additional mutex / boolean?

If `cancel` is called first, then I think that might still be an issue (?). That could be solved maybe by adding a check if cancelling in `succeeded_current` to process that instead of `current_handle_->succeed(result);`?

I _believe_ that would work and involve no new variables / locks. What do you think?

Reply:
Actually, at first, I used this way too. But I found an issue, if I was not mistaken.

> If succeeded_current is called first, then the update_mutex_ would be locked, and once it was done, then handle would no longer be active (or valid) and then it could use this condition to REJECT without the need of an additional mutex / boolean?

In this case, I think it should work. 

> If cancel is called first, then I think that might still be an issue (?). That could be solved maybe by adding a check if cancelling in succeeded_current to process that instead of current_handle_->succeed(result);?

But in this case. I think it was still an issue. Actually the goal state would be transformed to `CANCELING` only after this line executed(https://github.com/ros2/rclcpp/blob/543a3c39c16fdeb3f0109d0e906105870f547fe1/rclcpp_action/include/rclcpp_action/server.hpp#L367). 
So once `cancel_handle` returned `ACCEPT`, then checked if goal state was cancelling, it would be `false`, then `current_handle_->succeed(result);` would still be executed.

It seems that it's missing some mutexes (or other synchronization mechanisms) between goal handle methods and goal handle callbacks.  If so, it might be an rclcpp issue. What's your opinion?

Reply:
You're definitely welcome to bring up those synchronization issues to rclcpp if you like, definitely would rather have them fixed upstream. 

I think that would work, what we'd do is in `succeeded_current`, after we lock the mutex (so that if cancel was called first, this function will be blocked until completed), **then** check if cancellation if requested to process that. If not, then do `succeeded(result)`. So it'd be an if/else to do `current_handle_->succeeded(result);` or `current_handle_->canceled(result);`.

Reply:
yes, I had already opened an issue in rclcpp project. If I have a better solution, I will submit a PR to rclcpp.

OK, this would be a solution about the second case if we treat this as a synchronization issue of rclcpp.

I think I had made a mistake for the first case cause these cases might be a little complicated.
If `succeeded_current` is called first, then `cancel_handle ` would be blocked. Then `succeeded_current` would never be completed. Because in rclcpp, once  `cancel_handle ` was called, rclcpp 's mutex (`goal_handles_mutex_`) would be locked. 
Both `current_handle_->succeeded(result);` and `current_handle_->canceled(result);` would try to lock the mutex too. So that it became a deadlock issue.
Should we treat this as a synchronization issue of rclcpp too?

Some mentioned code here:
https://github.com/ros2/rclcpp/blob/47f21dab3d2c01710c8ea04784e970eec4a92e15/rclcpp_action/include/rclcpp_action/server.hpp#L355
https://github.com/ros2/rclcpp/blob/47f21dab3d2c01710c8ea04784e970eec4a92e15/rclcpp_action/include/rclcpp_action/server.hpp#L393
https://github.com/ros2/rclcpp/blob/47f21dab3d2c01710c8ea04784e970eec4a92e15/rclcpp_action/include/rclcpp_action/server_goal_handle.hpp#L194
https://github.com/ros2/rclcpp/blob/47f21dab3d2c01710c8ea04784e970eec4a92e15/rclcpp_action/include/rclcpp_action/server_goal_handle.hpp#L214

Reply:
> If succeeded_current is called first, then cancel_handle would be blocked. Then succeeded_current would never be completed.

Why? Cancel would be blocked until succeeded current was completed, that's true, but there's no reason it wouldn't be completed that I can see. Handling of the cancel request would entirely be blocked until succeeded_current is done and once its done then the cancel callback would see its not active anymore (from `!handle->is_active()` you added) and return.

Reply:
> Cancel would be blocked until succeeded current was completed, that's true.

Yes. To navigation2, I think it's a correct logic.

> Handling of the cancel request would entirely be blocked until succeeded_current is done.

It would be blocked when doing `current_handle_->succeed(result);`, because it was trying to lock the rclcpp 's mutex (`goal_handles_mutex_`) which had been locked before `cancel_handle` called.

I think the problem might be that rclcpp has not considered about the user mutex locked in the callback.

Reply:
OK, I'd explain that more concisely in the rclcpp issue: cancel/succeed are problematic with a race condition when the user **has** to lock the goal handle resource due to the multithreaded implications of the action server.

Reply:
Do you see any way around this other than throwing in more mutexes in this? Is there a solution in actions in rclcpp that would resolve this issue?

Reply:
https://github.com/KavenYau/rclcpp/commit/d1485ce6cfe46834b01585a73495b8c14bc4615e

I hope this patch could fix the deadlock issue. I'll test it in the next few days.

Reply:
I have tested by adding a unit test (https://github.com/KavenYau/rclcpp/commit/9fe631a20e6cb22a1db5812c0ae662b6a45b9d3c).

Without the this patch KavenYau/rclcpp@b8f567b, the test was failed.
After applying the patch, the test was passed.

Reply:
Is that an open PR on rclcpp? I'd support that change!

Reply:
Yes, this is the PR. https://github.com/ros2/rclcpp/pull/1635

Reply:
Is our current status for this now to wait until that's put in or anything we can do here in the meantime?

Reply:
The initial patch which I submited might be a temporary solution.

But I think the final solution should be based on rclcpp's changes. Maybe we can do some changes here first?